### DISCUSSION TITLE
Airflow scheduler with Kubernetes executor trying to adopt pod from other deployment

### DISCUSSION THREAD

**Discussion by bparhy** [apache/airflow#20219] [General]
### Apache Airflow version

2.1.3

### What happened

We are using Airflow version 2.1.3
Using kubernetes executor
We have other deployment in this k8s namespace which are running using airflow 1.10.10 and 1.10.12 etc.
For the airflow version 2.1.3 we are seeing an error in the scheduler logs where it looks like the scheduler is trying to adopt pods which are scheduled by other airflow deployments. Below is the error.

Failed to adopt pod <pod_id> Reason: (422)

Reason: Unprocessable Entity
HTTP response headers:

This mainly happens for pods/task which are completed already.

### What you expected to happen

Airflow scheduler should not attempt to adopt pods from other deployments.

### How to reproduce

Create two airflow deployments one with version 1.10.10 and other with 2.1.3

Set up a dag in 1.10.10 with a frequency of 5 minutes and then check the logs for airflow version 2.1.3.

You should see 2.1.3 scheduler trying to adopt pods with the below error

Failed to adopt pod <pod_id> Reason: (422)

Reason: Unprocessable Entity
HTTP response headers:

### Operating System

linux

### Versions of Apache Airflow Providers

Not dependent on provider version

### Deployment

Other 3rd-party Helm chart

### Deployment details

We are deploying on k8s instance with 2 scheduler and 1 web server with airflow 2.1.3
We are using helm to deploy and metadata DB is postgres with AWS Aurora.

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by potiuk** [comment]
You should make sure each airflow deployment is in a different namespace if I am not mistaken @jedcunningham @kaxil ? Can you please confirm it?

**Comment by bparhy** [comment]
Is this a new requirement as all our previous deployment prior to 2.1.3 is running without this issue. Also I am seeing this fix https://github.com/apache/airflow/pull/14795 is that related ? 
Thanks for responding so quick.

**Comment by potiuk** [comment]
> Is this a new requirement as all our previous deployment prior to 2.1.3 is running without this issue. Also I am seeing this fix #14795 is that related ? Thanks for responding so quick.

I am not sure - it's just seems logical to me to keep each airlfow deployment in a separate namespace (that 's what I'd do in general at least). But yeah. I might be wrong and Airlfow should work this way (but it also could be accidental that it worked)

**Comment by potiuk** [comment]
That's why I am asking those who **likely** know better :)

**Comment by jedcunningham** [comment]
Right now, the most reliable way is to run in separate namespaces.

Both running multiple instances in a single namespace and using the `multi_namespace_mode` option have various edge cases that just aren't handled currently (though, it is on my radar!). This hasn't changed in recent versions and it's likely you just got lucky or didn't notice issues previously.

**Comment by potiuk** [comment]
Yeah. My though exactly

**Comment by potiuk** [comment]
Moving that into a discussion then. I think, if we want to make multiple airflows in one namespace that shoudl be a separate feature. So maybe @bparhy - if you really think this is needed and you cannot rearrange your airflows to multipe instances - opening a feature request might be a good thing,

**Comment by devender-yadav** [comment]
I am testing with 2.2.3 and it's failing in [_adopt_completed_pods](https://github.com/apache/airflow/blob/2.2.3/airflow/executors/kubernetes_executor.py#L723) because I guess it's trying to set - `pod.metadata.labels['airflow-worker'] = '' ` for completed pods via 1.10.10 deployment in the same namespace.

**Reply by devender-yadav** [reply]
@jedcunningham what do you think about handling only subset of pods filtered using `queued_by_job_id` or `job_id` in [adopt_completed_pods](https://github.com/apache/airflow/blob/2.2.3/airflow/executors/kubernetes_executor.py#L686)?

**Comment by nsarvesh26** [comment]
For some reason this is happening even with single airflow deployment. Tried with 2.1.3 and 2.2.2, the issue is same. My tasks are getting failed whenever scheduler get restarted. I have deployed airflow in default namespace.

**Reply by olivermeyer** [reply]
I opened an [issue](https://github.com/apache/airflow/issues/23885) for this. I'm running 2.2.4 and running pods are consistently killed when the scheduler restarts.
